src/libraries/System.Linq/src/System/Linq/Average.cs

```diff
@@ -122,360 +55,130 @@ public static double Average(this IEnumerable<long> source)
 
-                double sum = e.Current;
+                TAccumulator sum = TAccumulator.CreateChecked(e.Current);
                 long count = 1;
                 while (e.MoveNext())
                 {
-                    // There is an opportunity to short-circuit here, in that if e.Current is
-                    // ever NaN then the result will always be NaN. Assuming that this case is
-                    // rare enough that not checking is the better approach generally.
-                    sum += e.Current;
-                    ++count;
+                    checked { sum += TAccumulator.CreateChecked(e.Current); }
+                    count++;
                 }
 
-                return sum / count;
+                return TResult.CreateChecked(sum) / TResult.CreateChecked(count);
             }
         }
 
-        public static double? Average(this IEnumerable<double?> source)
-        {
-            if (source == null)
-            {
-                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
-            }
 
-            using (IEnumerator<double?> e = source.GetEnumerator())
-            {
-                while (e.MoveNext())
-                {
-                    double? v = e.Current;
-                    if (v.HasValue)
-                    {
-                        double sum = v.GetValueOrDefault();
-                        long count = 1;
-                        checked
-                        {
-                            while (e.MoveNext())
-                            {
-                                v = e.Current;
-                                if (v.HasValue)
-                                {
-                                    sum += v.GetValueOrDefault();
-                                    ++count;
-                                }
-                            }
-                        }
-
-                        return sum / count;
-                    }
-                }
-            }
-
-            return null;
-        }
-
-        public static decimal Average(this IEnumerable<decimal> source)
-        {
-            if (source.TryGetSpan(out ReadOnlySpan<decimal> span))
-            {
-                if (span.IsEmpty)
-                {
-                    ThrowHelper.ThrowNoElementsException();
-                }
+        public static double? Average(this IEnumerable<int?> source) => Average<int, long, double>(source);
 
-                return Sum(span) / span.Length;
-            }
+        public static double? Average(this IEnumerable<long?> source) => Average<long, long, double>(source);
 
-            using (IEnumerator<decimal> e = source.GetEnumerator())
-            {
-                if (!e.MoveNext())
-                {
-                    ThrowHelper.ThrowNoElementsException();
-                }
+        public static float? Average(this IEnumerable<float?> source) => Average<float, double, double>(source) is double result ? (float)result : null;
```

Comment:
Is something like this needed for the ` Sum(this IEnumerable<float?> source, ...)` operations too?

Reply:
I don't think so.  This one needs it because the TResult is different from TSource, and for float we want TResult to be double to match the precision that was there before, so the callee needs to do the cast (or else we'd need yet another generic type parameter).